source/_posts/2018-01-02-Getting-started-with-Vapor-3.md

```diff
@@ -0,0 +1,344 @@
+    typealias Database = MySQLDatabase
+    typealias ID = Int
+    static let idKey = \Post.id
+}
+```
+
+The only steps we need to take in order to make our `Post` become a `Model` is to define the database we're using for this model, define how our models are uniquely identified and define what the keypath to our id field is.
+
+#### Migrations
+
+```swift
+extension Post: Migration {}
+```
+
+That's pretty much it. Using `Codable` Vapor is now able to convert your Swift types into database field types. To make sure your migration is being run you'll have to use a migration configuration in your `configure.swift`:
+
+```swift
+var migrationConfig = MigrationConfig()
+migrationConfig.add(model: Post.self, database: .mysql)
+services.use(migrationConfig)
+```
+
+This will obviously not be a solution to all cases, and most of the time you will probably have specific requirements for your db fields. One example could be to lower the length of your VARCHAR since a `String` in Swift will be turned into a 255 characters long VARCHAR, which might not be what you want if you want [to support emojis](https://github.com/nodes-vapor/readme/blob/master/Documentation/how-to-support-emojis.md). To handle your migration manually, you can instead specify the `prepare` function yourself instead of using the default behaviour:
+
+```swift
+extension Post: Migration {
+    static func prepare(on connection: MySQLConnection) -> Future<Void> {
+       return MySQLDatabase.create(self, on: connection) { builder in
+          try builder.field(for: \.id)
+          try builder.field(for: \.title)
+          try builder.field(type: .varChar(length: 191), for: \.body)
+        }
+    }
+}
+```
+
+### Routing
+
+Let's go through some of common use cases for dealing with a `Model`:
+
+- Retrieving all instances of a model.
+- Retriving one specific instance of a model using a unique identifier.
+- Creating a new instance of a model.
+- Updating an instance of a model.
+- Deleting an instance of a model.
+
+Before setting up the functionality for these different cases, we need to do two things first. First let's make our database available through our request (this could e.g. be added to our `configure.swift`):
+
+```swift
+extension Request: DatabaseConnectable {}
+```
+
+This will make it a bit easier to interact with the database by using the request as we will see soon.
+
+Next, we want to make our `Post` model conform to `Parameter` so that it can be used as a type-safe parameter for our routes. Go ahead and add this to `Post.swift`:
+
+```swift
+extension Post: Parameter {}
+```
+
+With that in place, we can now go through the different cases for interacting with our model type. Before we add routes for each case, let's go ahead and a `RouteGroup` to our `routes.swift` which is a `RouteCollection` where we are supposed to register our routes:
+
+```swift
+func boot(router: Router) throws {
+    // Posts
+    RouteGroup(cascadingTo: router).group("posts") { group in
```

Comment:
```swift
let posts = router.grouped("posts")
posts.get(...) // etc
```
To me, it seems that "Controllers" are an unnecessary concept. You can just provide functions. Dependency injection is not really supported anyway because of the async nature of Vapor 3. Within the functions you have the get the dependencies from the request.